src/ast2ram/ClauseTranslator.h

```diff
@@ -47,38 +48,50 @@ class Statement;
 
 namespace souffle::ast2ram {
 
+class AstToRamTranslator;
 class TranslatorContext;
 class ValueIndex;
 
 class ClauseTranslator {
 public:
-    ClauseTranslator(const TranslatorContext& context, SymbolTable& symbolTable)
-            : context(context), symbolTable(symbolTable) {}
+    ClauseTranslator(const TranslatorContext& context, SymbolTable& symbolTable);
+    ~ClauseTranslator();
 
-    /** Generate RAM code for a clause */
-    Own<ram::Statement> translateClause(
-            const ast::Clause& clause, const ast::Clause& originalClause, int version = 0);
+    /** Entry points */
+    static Own<ram::Statement> translateNonRecursiveClause(
+            const TranslatorContext& context, SymbolTable& symbolTable, const ast::Clause& clause);
+
+    static VecOwn<ram::Statement> translateRecursiveClause(const TranslatorContext& context,
+            SymbolTable& symbolTable, const ast::Clause* clause, const std::set<const ast::Relation*>& scc);
 
 protected:
     const TranslatorContext& context;
     SymbolTable& symbolTable;
 
-    // value index to keep track of references in the loop nest
-    Own<ValueIndex> valueIndex = mk<ValueIndex>();
+    size_t version{0};
+    std::vector<ast::Atom*> sccAtoms{};
 
-    virtual Own<ram::Operation> createProjection(
-            const ast::Clause& clause, const ast::Clause& originalClause) const;
-    virtual Own<ram::Condition> createCondition(const ast::Clause& originalClause) const;
+    bool isRecursive() const;
 
-private:
-    std::vector<const ast::Argument*> generators;
-    std::vector<const ast::Node*> operators;
+    std::string getClauseAtomName(const ast::Clause& clause, const ast::Atom* atom) const;
 
-    Own<ast::Clause> getReorderedClause(const ast::Clause& clause, const int version) const;
+    virtual Own<ram::Operation> addNegatedAtom(Own<ram::Operation> op, const ast::Atom* atom) const;
+    virtual Own<ram::Operation> addNegatedDeltaAtom(Own<ram::Operation> op, const ast::Atom* atom) const;
 
-    /** Operation levelling */
-    int addGeneratorLevel(const ast::Argument* arg);
-    int addOperatorLevel(const ast::Node* node);
+    Own<ValueIndex> valueIndex;
+
+    Own<ram::Statement> translateClause(const ast::Clause& clause);
+
+    /** Main clause translation */
+    virtual Own<ram::Statement> createRamFactQuery(const ast::Clause& clause) const;
+    virtual Own<ram::Statement> createRamRuleQuery(const ast::Clause& clause);
```

Comment:
Is the variations with virtual and const in the new methods intentional?

Reply:
Yep. `createRamFactQuery` doesn't need to index the clause, while `createRamRuleQuery` does. Indexing is a state-changing function. If it becomes more natural (and there's time), I might just pass around the index as needed.